internal/controllers/machineset/machineset_controller.go

```diff
@@ -966,6 +994,146 @@ func (r *Reconciler) getMachineNode(ctx context.Context, cluster *clusterv1.Clus
+			return result, nil
+		}
+	}
+
+	// All pre-flight checks passed.
+	return ctrl.Result{}, nil
+}
+
+func (r *Reconciler) controlPlaneStablePreflightCheck(ctx context.Context, controlPlane *unstructured.Unstructured) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	cpKlogRef := klog.KRef(controlPlane.GetNamespace(), controlPlane.GetName())
+
+	// Check that the control plane is not provisioning.
+	isProvisioning, err := contract.ControlPlane().IsProvisioning(controlPlane)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform control plane pre-flight check: failed to check if ControlPlane %s is provisioning", cpKlogRef)
+	}
+	if isProvisioning {
+		log.Info(fmt.Sprintf("Pre-flight check failed: ControlPlane %s is provisioning", cpKlogRef))
+		return ctrl.Result{RequeueAfter: preflightFailedRequeueAfter}, nil
+	}
+
+	// Check that the control plane is not upgrading.
+	isUpgrading, err := contract.ControlPlane().IsUpgrading(controlPlane)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform control plane pre-flight check: failed to check if the ControlPlane %s is upgrading", cpKlogRef)
+	}
+	if isUpgrading {
+		log.Info(fmt.Sprintf("Pre-flight check failed: ControlPlane %s is upgrading", cpKlogRef))
+		return ctrl.Result{RequeueAfter: preflightFailedRequeueAfter}, nil
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *Reconciler) kubernetesVersionPreflightCheck(ctx context.Context, cpVersion string, ms *clusterv1.MachineSet) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	// If the version is not defined on the MachineSet then we cannot check for the version skew policies.
+	// Return early.
+	if ms.Spec.Template.Spec.Version == nil {
+		return ctrl.Result{}, nil
+	}
+	msVersion := *ms.Spec.Template.Spec.Version
+
+	cpSemver, err := semver.ParseTolerant(cpVersion)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform kubernetes version skew pre-flight check: failed to parse ControlPlane version %s", cpVersion)
+	}
+	msSemver, err := semver.ParseTolerant(msVersion)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform kubernetes version skew pre-flight check: failed to parse MachineSet version %s", msVersion)
+	}
+
+	// Check the Kubernetes version skew policy.
+	// => MS minor version cannot be greater than the Control Plane minor version.
+	// => MS minor version cannot be older than 2 minor version of Control Plane.
+	// Kubernetes skew policy: https://kubernetes.io/releases/version-skew-policy/#kubelet
+	floorVersion := semver.Version{
+		Major: cpSemver.Major,
+		Minor: cpSemver.Minor - 2,
+		Patch: 0,
+	}
+	if msSemver.GT(cpSemver) || msSemver.LT(floorVersion) {
+		log.Info(fmt.Sprintf("Pre-flight check failed: MachineSet version (%s) and ControlPlane version (%s) do not conform to the kubernetes version skew policy", msVersion, cpVersion))
+		return ctrl.Result{RequeueAfter: preflightFailedRequeueAfter}, nil
+	}
```

Comment:
Should this be in a validation webhook?

Reply:
Which one? MD + MS?

At least for the latter check, the webhook is not enough as the CP version can just change and then existing MachineSets "fall" outside of the policy

Reply:
Maybe we want to have it additionally in the webhook?

Reply:
Could be on both sides?

Reply:
I'll have to think a bit about it, but sounds reasonable at a first glance

Reply:
I think it's fine to add both checks to the MD + MS validating webhook.

But they should only block on create and on update only if the version is changing. They shouldn't block an update where the version is not changed in my opinion

I would prefer implementing the webhooks in a follow-up PR though.

Hm, but it wouldn't make sense to disable the validation on the webhooks via the skip preflight check annotation. As the webhook is not a preflight check. We could maybe choose to only return warnings in the webhook?

@vincepri WDYT?